### PR TITLE
Dependent Bindings

### LINKED ISSUE #136: Top sort bindings before wrapping them over an expression
Extracted from https://github.com/rexim/Morganey/issues/109#issuecomment-241452378

### PR DESCRIPTION
**Pull request by rexim**
This PR significantly changes the semantic of Morganey for good. Now dependent bindings work correctly. Also now using of non-existing bindings and binding loop generate an error.

---

Close #136

### CODE REVIEW

File: kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala [hunk 1]
```diff
@@ -46,6 +47,30 @@ sealed trait LambdaTerm extends MorganeyNode {
...
+
+  def addDependentBindings(context: Seq[MorganeyBinding],
+                           currentVars: Set[String] = Set()): Either[String, LambdaTerm] = {
+    val conflictedVars = freeVars & currentVars
+
+    if (conflictedVars.isEmpty) {
+      val bindings: List[Either[String, MorganeyBinding]] = freeVars.toList.map { x =>
+        context
+          .find(_.variable.name == x)
+          .toRight(s"Unexisting binding: $x")
```
Comment (Minoru) [inline kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala]: If you meant "a binding that doesn't exist", then you should've used "non-existent", not "unexisting".

File: kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala [hunk 2]
```diff
@@ -46,6 +47,30 @@ sealed trait LambdaTerm extends MorganeyNode {
       case (MorganeyBinding(variable, value), acc) =>
         LambdaApp(LambdaFunc(variable, acc), value)
     }
+
+  def addDependentBindings(context: Seq[MorganeyBinding],
+                           currentVars: Set[String] = Set()): Either[String, LambdaTerm] = {
+    val conflictedVars = freeVars & currentVars
```
Comment (rexim) [inline kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala]: `conflictingVars` is better I think

File: src/test/scala/me/rexim/morganey/AutocompletionSpec.scala [hunk 3]
```diff
@@ -16,7 +16,7 @@ class AutocompletionSpec extends FlatSpec with Matchers with TestTerms  {
   def autocomplete(line: String, cursor: Int, knownNames: List[String]): Set[CharSequence] = {
     val id = I(lvar("x"))
     val fakeBindings = knownNames.map(name => MorganeyBinding(lvar(name), id))
-    val moduleFinder = new ModuleFinder(List(new File("./std/")))
+    val moduleFinder = new ModuleFinder(List(new File("./test-data/load-autocomplete/")))
```
Comment (keddelzz) [inline src/test/scala/me/rexim/morganey/AutocompletionSpec.scala]: Very nice idea to use a different folder for the autocompletion-tests.

Aren't test resources in a project with maven directory structure located at `src/test/resources`?

Comment (rexim) [reply src/test/scala/me/rexim/morganey/AutocompletionSpec.scala]: Good point! It'd be perfect place for those files! I'll try to move them there.

Comment (rexim) [reply src/test/scala/me/rexim/morganey/AutocompletionSpec.scala]: Which actually brings up another interesting question: do we want the std-lib to be resource for the uberjar? #45 Which is indeed a separate question and can be discussed in #45 if needed. Simply wanted to mentioned it here.

**Comment by rexim** [comment]
@keddelzz @ForNeVeR  this PR is finally cooked! Please take a look when you get a chance. :)

File: kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala [hunk 4]
```diff
@@ -10,11 +12,36 @@ sealed trait LambdaTerm extends MorganeyNode {
...
       case (MorganeyBinding(variable, value), acc) =>
         LambdaApp(LambdaFunc(variable, acc), value)
     }
+
+  private def addBindingsImpl(context: Seq[MorganeyBinding],
+                      currentVars: List[String]): Either[BindingError, LambdaTerm] = {
+      val bindings = freeVars.toList.map { x =>
+        if (!currentVars.contains(x)) {
+          context
+            .find(_.variable.name == x)
```
Comment (ForNeVeR) [inline kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala:26]: This gonna become extremely inefficient with increasing amount of variables in context. Probably construct a local map in this function?

Comment (rexim) [reply kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala:26]: Yep, I'm aware of that. I wanted to create a separate issue for that, but forgot. Thanks for reminding me!

Comment (rexim) [reply kernel/src/main/scala/me/rexim/morganey/ast/LambdaTerm.scala:26]: Done

**Comment by ForNeVeR** [comment]
LGTM 👍

**Pull request merged**